### ISSUE TITLE
Infinate hang when Elasticsearch restarts

### ISSUE THREAD

**Issue by jordansissel** [logstash-plugins/logstash-output-elasticsearch#142]
(This issue was originally filed by @mdconner at https://github.com/elastic/logstash/issues/1655)

---

Logstash appears to go into infinite hang after events are processed after Elasticsearch restarts.  
Tested with Logstash 1.4.1 & Elasticsearch 1.1.1 on RHEL 6.5 by the following steps (tested with File input also):
1. Start Elasticsearch from command line and wait for "started" message
2. Start Logstash from command line as agent with following config file
   input {  stdin { } } 
   output {   stdout {  }
     elasticsearch { cluster => "elasticsearch"  }
   }
3. Send test events (stdin)
4. Stop Elasticsearch
5. Send test events (stdin)
6. Start Elasticsearch
7. Send test events (stdin)

Seems to recover successfully when no events sent before Elasticsearch restarts (step 5).  Here is the entire console output (Exception occurs after Test 5):

```
[root@lnx65 bin]# ./logstash agent --verbose -f ../mdc_stdin.conf
Pipeline started {:level=>:info}
log4j, [2014-08-21T20:10:44.283]  INFO: org.elasticsearch.node: [logstash-lnx65.localdomain-17807-2010] version[1.1.1], pid[17807], build[f1585f0/2014-04-16T14:27:12Z]
log4j, [2014-08-21T20:10:44.283]  INFO: org.elasticsearch.node: [logstash-lnx65.localdomain-17807-2010] initializing ...
log4j, [2014-08-21T20:10:44.316]  INFO: org.elasticsearch.plugins: [logstash-lnx65.localdomain-17807-2010] loaded [], sites []
log4j, [2014-08-21T20:10:47.029]  INFO: org.elasticsearch.node: [logstash-lnx65.localdomain-17807-2010] initialized
log4j, [2014-08-21T20:10:47.029]  INFO: org.elasticsearch.node: [logstash-lnx65.localdomain-17807-2010] starting ...
log4j, [2014-08-21T20:10:47.141]  INFO: org.elasticsearch.transport: [logstash-lnx65.localdomain-17807-2010] bound_address {inet[/0:0:0:0:0:0:0:0:9301]}, publish_address {inet[/10.0.2.15:9301]}
log4j, [2014-08-21T20:10:50.358]  INFO: org.elasticsearch.cluster.service: [logstash-lnx65.localdomain-17807-2010] detected_master [lnx65][El3W2hxoQvqkvIWo0fT4sg][lnx65.localdomain][inet[/10.0.2.15:9300]], added {[lnx65][El3W2hxoQvqkvIWo0fT4sg][lnx65.localdomain][inet[/10.0.2.15:9300]],}, reason: zen-disco-receive(from master [[lnx65][El3W2hxoQvqkvIWo0fT4sg][lnx65.localdomain][inet[/10.0.2.15:9300]]])
log4j, [2014-08-21T20:10:50.368]  INFO: org.elasticsearch.discovery: [logstash-lnx65.localdomain-17807-2010] elasticsearch/SeDbmuOZTlq5DxxCnENasQ
log4j, [2014-08-21T20:10:50.373]  INFO: org.elasticsearch.node: [logstash-lnx65.localdomain-17807-2010] started
New Elasticsearch output {:cluster=>"elasticsearch", :host=>nil, :port=>"9300-9305", :embedded=>false, :protocol=>"node", :level=>:info}
Automatic template management enabled {:manage_template=>"true", :level=>:info}
Using mapping template {:template=>"{  \"template\" : \"logstash-*\",  \"settings\" : {    \"index.refresh_interval\" : \"5s\"  },  \"mappings\" : {    \"_default_\" : {       \"_all\" : {\"enabled\" : true},       \"dynamic_templates\" : [ {         \"string_fields\" : {           \"match\" : \"*\",           \"match_mapping_type\" : \"string\",           \"mapping\" : {             \"type\" : \"string\", \"index\" : \"analyzed\", \"omit_norms\" : true,               \"fields\" : {                 \"raw\" : {\"type\": \"string\", \"index\" : \"not_analyzed\", \"ignore_above\" : 256}               }           }         }       } ],       \"properties\" : {         \"@version\": { \"type\": \"string\", \"index\": \"not_analyzed\" },         \"geoip\"  : {           \"type\" : \"object\",             \"dynamic\": true,             \"path\": \"full\",             \"properties\" : {               \"location\" : { \"type\" : \"geo_point\" }             }         }       }    }  }}", :level=>:info}

2014-08-21T18:10:53.394+0000 lnx65.localdomain 
Test 1
2014-08-21T18:10:58.896+0000 lnx65.localdomain Test 1
Test 2
2014-08-21T18:11:01.376+0000 lnx65.localdomain Test 2
log4j, [2014-08-21T20:11:06.351]  INFO: org.elasticsearch.discovery.zen: [logstash-lnx65.localdomain-17807-2010] master_left [[lnx65][El3W2hxoQvqkvIWo0fT4sg][lnx65.localdomain][inet[/10.0.2.15:9300]]], reason [transport disconnected (with verified connect)]
log4j, [2014-08-21T20:11:06.353]  WARN: org.elasticsearch.discovery.zen: [logstash-lnx65.localdomain-17807-2010] master_left and no other node elected to become master, current nodes: {[logstash-lnx65.localdomain-17807-2010][SeDbmuOZTlq5DxxCnENasQ][lnx65.localdomain][inet[/10.0.2.15:9301]]{client=true, data=false},}
log4j, [2014-08-21T20:11:06.354]  INFO: org.elasticsearch.cluster.service: [logstash-lnx65.localdomain-17807-2010] removed {[lnx65][El3W2hxoQvqkvIWo0fT4sg][lnx65.localdomain][inet[/10.0.2.15:9300]],}, reason: zen-disco-master_failed ([lnx65][El3W2hxoQvqkvIWo0fT4sg][lnx65.localdomain][inet[/10.0.2.15:9300]])

2014-08-21T18:11:14.608+0000 lnx65.localdomain 

Test 3
Test 4
Test 5
Exception in thread "elasticsearch[logstash-lnx65.localdomain-17807-2010][generic][T#2]" org.elasticsearch.cluster.block.ClusterBlockException: blocked by: [SERVICE_UNAVAILABLE/1/state not recovered / initialized];[SERVICE_UNAVAILABLE/2/no master];
	at org.elasticsearch.cluster.block.ClusterBlocks.globalBlockedException(ClusterBlocks.java:138)
	at org.elasticsearch.cluster.block.ClusterBlocks.globalBlockedRaiseException(ClusterBlocks.java:128)
	at org.elasticsearch.action.bulk.TransportBulkAction.executeBulk(TransportBulkAction.java:197)
	at org.elasticsearch.action.bulk.TransportBulkAction.access$000(TransportBulkAction.java:65)
	at org.elasticsearch.action.bulk.TransportBulkAction$1.onFailure(TransportBulkAction.java:143)
	at org.elasticsearch.action.support.TransportAction$ThreadedActionListener$2.run(TransportAction.java:117)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:744)
log4j, [2014-08-21T20:11:45.493]  INFO: org.elasticsearch.cluster.service: [logstash-lnx65.localdomain-17807-2010] detected_master [lnx65][7vpAU211RHemHzNw1u9EyQ][lnx65.localdomain][inet[/10.0.2.15:9300]], added {[lnx65][7vpAU211RHemHzNw1u9EyQ][lnx65.localdomain][inet[/10.0.2.15:9300]],}, reason: zen-disco-receive(from master [[lnx65][7vpAU211RHemHzNw1u9EyQ][lnx65.localdomain][inet[/10.0.2.15:9300]]])

Test 6
Test 7
```

**Comment by titbarros** [comment]
Any news o  this issue? Is there any work around? Maybe a Way to  check logstash process status and restart it

**Comment by skikd636** [comment]
My work around my not work for you since I have a rabbitmq server running in the middle of logstash_forwarder and the actual logstash that processes the records into ES.  I have a simple process that fires ever 30 minutes that checks my queues to see if they are getting backed up.  If they are I restart the LS process that is hung and it will catch up with that 30 minutes since I'm not running the system at max capacity.  Been working for me until this can get resolved.

**Comment by jordansissel** [comment]
We are investigating this. No news right now, but hopefully soon.

**Comment by jordansissel** [comment]
For folks experiencing this, Does Logstash never recover from this?

**Comment by skikd636** [comment]
For me logstash never comes back it will just spin until it is restarted.

**Comment by jordansissel** [comment]
@skikd636 thanks for the info! <3

**Comment by kerryjj** [comment]
We have what looks like the same issue with logstash 1.5.0.

We are running elasticsearch 1.4.5, but in our case it wasn't elasticsearch restarting. There was some network error that caused logstash to fail to connect to elasticsearch temporarily, maybe because the two servers are in different availability zones in AWS.

Either way, it looks like once logstash can't contact elasticsearch it hangs and fails to reconnect.

**Comment by vksinghh25** [comment]
Are there any updates on this issue? 

I now have an ELK setup with Kafka that works well except for this issue. The logstash process hangs anytime there is an error connecting to the ES cluster. Even restarting logstash doesn't seem to always fix the problem for me.

**Comment by jordansissel** [comment]
> Even restarting logstash doesn't seem to always fix 

This is different than other reported workarounds. As far as I know, users who hit this are able to restart Logstash and the problem resolves. Recommend you open another ticket and provide any logs (logstash --debug logs), elasticsearch logs at the time, your logstash config, etc.

**Comment by JosBergman** [comment]
Any news concerning this issue? We've experienced this one also last friday..

**Comment by shang1989** [comment]
I am also seeing this issue. logstash 1.4.. Expect solutions

**Comment by tkrajca** [comment]
Any news on this one? We also have the same issue. logstash 1.4, ES 1.1

**Comment by untergeek** [comment]
@tkrajca Logstash 1.4 is EOL and will not be developed further, as is Elasticsearch 1.1 and 1.2.

Any fix for this will be in a newer release.

**Comment by tkrajca** [comment]
I am happy to update, I just used the defaults for CentOS 6.5
